### ISSUE TITLE
Expose beforeAll/beforeEach/afterAll/afterEach in require'd files

### ISSUE THREAD

**Issue by keithamus** [mochajs/mocha#764]
Part of the utility of having mocha require in a common .js file (`--require`) is to be able to set up some common convenience features for ALL tests.

It would be useful to have some hooks similar to, or exactly, the /(after|before)(All|Each)/ hooks.

**Use case**: Being able to integrate tools like sinon more deeply with mocha, allowing automatic tear-downs of stubs/mocks for each/all steps.

**Comment by rstacruz** [comment]
what sort of feedback is still needed here? would love to see this. at the moment, i have to use `require('./setup')` on top of all my test files.

**Comment by pigulla** [comment]
Actually, there's another situation where this would be really useful.

I'm using [Wallaby](http://wallabyjs.com/docs/index.html) to run my tests in parallel. For this to work with the integration tests I've to use a separate database per test runner (I simply append the process id to the database name). This is done in `--require setup.js` file and it works perfectly.

The issue I'm having is: how can I drop that temporary database when the test is complete? I can't use `after()` in my `setup.js` because that hook is not exposed there. And I can't use the "root suite before/after" solution [proposed here](https://github.com/mochajs/mocha/issues/1460) either because I have no control over which test files are executed in which Wallaby runner (I'd need the spec with the global hooks to be run in _all_ Wallaby workers).

**Comment by eagleeye** [comment]
>I still don't understand why this is necessary

I have a lot of use-cases of this feature.
- Globally start selenium in `before` and shutdown in `after`
- clean database `after` tests will run
- start web-services `before` tests
- build/compile library before test

**Comment by boneskull** [comment]
@eagleeye thanks for some good specifics

**Comment by jmeas** [comment]
It's a bummer that this isn't supported (easily) by the gulp CLI. I've been spoiled by gulp-mocha, which lets you do...

```js
gulp.src(['path/to/setup/*', 'path/to/unit/**/*'])
  .pipe(...)
```

Aside from changing the behavior of the -r flag, I prefer the API of Gulp. Although the syntax isn't as nice, I've recreated that behavior using the CLI by changing my glob:

```
{test/unit/setup/*,test/unit/specs/**/*}
```

I figured I'd post the option in case anyone else likes it more than the other workarounds suggested above :v:

**Comment by tbranyen** [comment]
This is exactly how I solved it as well in my tests. Removed the `--require` and moved the setup script to be the first test executed. Won't work for everyone's use case, but works fine for mine:

``` bash
mocha test/client/setup test/client/some-test
```

**Comment by evgeny-myasishchev** [comment]
The issue with ```mocha test/setup test/some-test``` approach is that you have to remember to include the ```test/setup``` if you need to run some individual test file (and I do that very often).

I have "test" script in package.json similar to this ```mocha test/setup.js test/*-spec.js``` and this is working well. But to run individual tests I have to always keep in mind to run them like this:

```mocha test/setup.js test/some-spec.js```

I found that adding "test/setup.js" to mocha.opts (as a last thing) is working well, not sure if it's by design, but it works with mocha 3.2.0

**Comment by tbranyen** [comment]
@evgeny-myasishchev if you only want to run one test, isn't that what `only` and `grep` are for?

**Comment by evgeny-myasishchev** [comment]
@tbranyen Yep, I'm using those options as well. For me it's quite often just faster and more flexible to use ```mocha test/some-spec.js [--watch --some-opts...]```

Also with only there is another issue if you use it with --watch. If you start some specs in a watch mode and "only" and then remove only, then mocha will "see" zero tests so I have to restart it.

**Comment by tbranyen** [comment]
@evgeny-myasishchev I hate that issue so much. If I could fix one thing about Mocha, it'd be that watch bug.

**Comment by glenjamin** [comment]
Looks like someone has nearly done this https://github.com/mochajs/mocha/pull/2544

**Comment by emahuni** [comment]
Ahhhhghhhhh!!! I am new to Mocha and according to the documentation this is supposed to be a cinch. I  thought that it should just work in required files. why isn't this working? I have read the comments but still don't get why you guys think that exposing these hooks is bad.

**Comment by ScottFreeCode** [comment]
For the record, Mocha should run files like this:

1. `--require <file>` from `mocha.opts` **without** `(before|after)(Each)?|describe|it` (*guaranteed* to run)
2. `--require <file>` from an NPM script **without** `(before|after)(Each)?|describe|it` (*guaranteed* to run *as long as run through that NPM script*)
3. `--require <file>` from commandline **without** `(before|after)(Each)?|describe|it` (whatever you specify on the CLI)
4. `<file>` (no `--require`) from `mocha.opts` **with** `(before|after)(Each)?|describe|it` (*guaranteed* to run)
5. `<file>` (no `--require`) from an NPM script **with** `(before|after)(Each)?|describe|it` (*guaranteed* to run *as long as run through that NPM script*)
6. `<file>` (no `--require`) from commandline **with** `(before|after)(Each)?|describe|it` (whatever you specify on the CLI)

For example, if you have six files named 1.js through 6.js that need to run from 1 to 6 with the first three not having `(before|after)(Each)?|describe|it` and the last three having `(before|after)(Each)?|describe|it`, you can do it with this in `mocha.opts`:

```txt
3.js
--require 1.js
```

...and this in `package.json`'s `scripts`:

```txt
"test": "mocha 5.js --require 2.js"
```

...and calling this on the CLI:

```txt
npm test -- 6.js --require 3.js
```

(Note that I've reversed the order of `--require`d files vs. plain test files relative to each other simply to emphasize that both their being `--require`d and the order of `mocha.opts` then `npm` script then CLI overrides it; individual file ordering only affects files of the same `--require`d/not-`--require`d type specified in the same one of these three places.)

*That should provide more than enough flexibility to satisfy all reasonable use cases*; as far as I can think of the only *possible* cases it doesn't cover (of which I'm not sure there is any value) are:

- running a file *without* `(before|after)(Each)?|describe|it` **after** one *with* `(before|after)(Each)?|describe|it`
- *guaranteeing* files of a given type (with vs without `(before|after)(Each)?|describe|it`) are run **after** those of the same (or the later) type specified on the CLI
- tooling that wraps Mocha or intercepts Mocha's behavior in a way that interferes with those points of control without providing a mechanism of its own to fully replace the behavior with which it interferes (this is a problem in any such tooling and should be addressed as an issue there)

It might be worth adding a small section to the documentation to highlight this relationship of `--require <file>` vs. `<file>` and `mocha.opts` vs commandline.

**Comment by machineghost** [comment]
I'm amazed this is still a problem.  It seems to me a pretty common use case to:

A) have a test setup file, which does thing like instantiating a Sinon sandbox

B) want to invoke mocha tests multiple ways; you might run `mocha *`, or you might run individual test files (some IDEs make this very easy), or you might have a testSuite.js file

C) not want to have to repeat boilerplate such as `import testSetup from 'testSetup'; beforeEach(testSetup);` in every test file

Given those (IMHO pretty sane) requirements, `--require` seems like the natural solution.  It ensures that no matter how you run your tests the proper setup happens first, without the need to repeat `beforeEach(testSetup)` in every file.

The only problem is, if you try to use `beforeEach` in a `--require` file, it won't exist.  Why, I have no idea, but it prevents the above solution from working.  You can't even bring in `beforeEach` manually with `import { beforeEach } from 'mocha'`; it will be `undefined`.

It would be _really_ convenient if Mocha just made the globals available to `--require` files, but until they do I found a workaround elsewhere:

   const doTestSetup = () => doWhatever();
   setTimeout(() => beforeEach(doTestSetup));

You can also use `defer` if you're a Lodash fan like myself.  By calling `beforeEach` inside the timeout function it actually will exist by the time that function executes, allowing you to do proper test setup via `--require`.

Hope this helps others.

**Comment by boneskull** [comment]
Ensure you've upgraded to a version of Mocha with `--file`, but here's an example:

```js
// test.spec.js
it('should be ok', function () {});
```
```js
// setup.js
afterEach(function () {
  console.log('a test happened');
});
```
```bash
$ mocha --file setup.js test.spec.js


  ✓ should be ok
a test happened

  1 passing (6ms)

```

**Comment by machineghost** [comment]
Just updated Mocha to be sure I wasn't being stupid :)

>+ mocha@3.5.3

Here's a real simple example:

**A.js**

window.foo();

**ATest.js**
import 'A';

**testSetup.js**
global.window = { foo: () => {});

`mocha --file testSetup ATest.js`

**Comment by boneskull** [comment]
Mocha doesn't support `import` in Node.js (and won't until no longer experimental)

**Comment by machineghost** [comment]
Oh, should have mentioned that I'm using Babel, so it's all getting transpiled down to `require` statements.  I'll update my example.

**Comment by boneskull** [comment]
Sorry, didn't see your Mocha version.  `--file` was added in v5.

**Comment by machineghost** [comment]
Ah!  I thought doing `npm i mocha` would get me the latest, but I guess I had something in my `package.json` blocking it.  In any case, I'm now at:

> + mocha@5.1.1

and when I run my example again I get ...

>(function (exports, require, module, __filename, __dirname) { window.foo();
>                                                              ^
>
?ReferenceError: window is not defined

**Comment by boneskull** [comment]
what does `mocha -v` output?

**Comment by machineghost** [comment]
> mocha -v
> error: unknown option `-v'

> mocha --version
> 3.5.3

WTF!  I just retried `npm -i mocha` again and again got that "+ mocha@5.1.1" ... but `mocha --version` still refers to 3.5.3!

Sorry for the confusion; not sure what's going on in my environment but it clearly doesn't appear to be Mocha's problem.  Thanks a bunch for all of the help.

**Comment by boneskull** [comment]
It's pretty simple; `mocha` is installed globally and is version 5.1.1.  This is what you get when you use `mocha`.  The `mocha` that's in your local `node_modules` is in `node_modules/.bin/mocha`.

**Comment by emahuni** [comment]
Can you pliz explain:
1. what co-mocha is doing that's enabling --require to work with global hooks?
2. if this is bad and why?
3. why mocha isn't implementing it like this

` mocha --require co-mocha test/app.mock.js test/**/*.test.js`

there is a `before()` in `app.mock.js` and it works well with the above but not with the below.

` mocha --require test/app.mock.js test/**/*.test.js`

Somehow `co-mocha` is exposing those hooks to a global scope enabling all requires to work with hooks. What got me back here was actually trying to figure out, how it's done. Actually it was a Google search and it got me back here, was surprised to see I had this same issue about a year ago. You guys should just expose the hooks, give an option to do so or explain this different relationship between `--file` and `--require` in the docs.

Please try to explain how mocha runs things, how it works with our tests in the documentation.
